salt/utils/openstack/neutron.py

```diff
@@ -53,22 +65,76 @@ class SaltNeutron(NeutronShell):
     Class for all neutronclient functions
     '''
 
-    def __init__(self, username, tenant_name, auth_url, password=None,
-                 region_name=None, service_type=None, **kwargs):
+    def __init__(
+        self,
+        username,
+        tenant_name,
+        auth_url,
+        password=None,
+        region_name=None,
+        os_auth_plugin=None,
+        use_keystoneauth=False,
+        **kwargs
+    ):
+
         '''
         Set up neutron credentials
         '''
         if not HAS_NEUTRON:
             return None
 
+        elif all([use_keystoneauth, HAS_KEYSTONEAUTH]):
+            self._new_init(username=username,
+                           project_name=tenant_name,
+                           auth_url=auth_url,
+                           region_name=region_name,
+                           os_auth_plugin=os_auth_plugin,
+                           password=password,
+                           **kwargs)
+        else:
+            self._old_init(username=username,
+                           tenant_name=tenant_name,
+                           auth_url=auth_url,
+                           region_name=region_name,
+                           os_auth_plugin=os_auth_plugin,
+                           password=password,
+                           **kwargs)
+
+    def _new_init(self, username, project_name, auth_url, region_name, password, os_auth_plugin, auth=None, verify=True, **kwargs):
+        if auth is None:
+            auth = {}
+
+        loader = keystoneauth1.loading.get_plugin_loader(os_auth_plugin or 'password')
+
+        self.client_kwargs = kwargs.copy()
+        self.kwargs = auth.copy()
+
+        self.kwargs['username'] = username
+        self.kwargs['project_name'] = project_name
+        self.kwargs['auth_url'] = auth_url
+        self.kwargs['password'] = password
+        if auth_url.endswith('3'):
+            self.kwargs['user_domain_name'] = kwargs.get('user_domain_name', 'default')
+            self.kwargs['project_domain_name'] = kwargs.get('project_domain_name', 'default')
+
+        self.client_kwargs['region_name'] = region_name
+        self.client_kwargs['service_type'] = 'network'
+
+        self.client_kwargs = sanitize_neutronclient(self.client_kwargs)
+        options = loader.load_from_options(**self.kwargs)
+        self.session = keystoneauth1.session.Session(auth=options, verify=verify)
+        self.network_conn = client.Client(session=self.session, **self.client_kwargs)
+
+    def _old_init(self, username, tenant_name, auth_url, region_name, password, os_auth_plugin, auth=None, verify=True, **kwargs):
         self.kwargs = kwargs.copy()
 
         self.kwargs['username'] = username
         self.kwargs['tenant_name'] = tenant_name
         self.kwargs['auth_url'] = auth_url
-        self.kwargs['service_type'] = service_type
+        self.kwargs['service_type'] = 'network'
```

Comment:
can't this be another name?  I would say we should default to 'network'  But i believe it is possible for it to be networkv2 on openstack instances, i think rackspace has networkv2.